### DISCUSSION TITLE
How to read localStorage in initCounterStore ？🤔

### DISCUSSION THREAD

**Discussion by lynn1286** [pmndrs/zustand#2630] [Q&A]
According to the documentation I'm using zustand in Nextjs and it's fine, but I'm thinking about a problem, I want to read the localStorage in the window object in initCounterStore, but unfortunately the server tells me ReferenceError: localStorage is not defined, but when I try to determine if localStorage exists I get an Error: Hydration failed because the initial UI does not match what was rendered on the server. and I'm stuck! 😖

```
// src/stores/counter-store.ts
import { createStore } from 'zustand/vanilla'

export type CounterState = {
  count: number
}

export type CounterActions = {
  decrementCount: () => void
  incrementCount: () => void
}

export type CounterStore = CounterState & CounterActions

export const initCounterStore = (): CounterState => {
  if(isServerRendering) return { count: new Date().getFullYear() }
 
 const count =
    localStorage.getItem('__free_report_open_banner__') ?? 0;
  return { count }
}

export const defaultInitState: CounterState = {
  count: 0,
}

export const createCounterStore = (
  initState: CounterState = defaultInitState,
) => {
  return createStore<CounterStore>()((set) => ({
    ...initState,
    decrementCount: () => set((state) => ({ count: state.count - 1 })),
    incrementCount: () => set((state) => ({ count: state.count + 1 })),
  }))
}
```

**Comment by dai-shi** [comment]
to avoid hydration mismatch, you first need to render with `0` and then read localStorage and update the value and render it again.

**Reply by lynn1286** [reply]
> to avoid hydration mismatch, you first need to render with `0` and then read localStorage and update the value and render it again.

That means I can't read local storage in the initCounterStore ? 
If I have a scenario where I want to get the value of local storage in initCounterStore and then go and call the interface to get the data and setup my store, is there any other way to accomplish this? Without breaking zustand
[Here's the code.](https://stackblitz.com/edit/nextjs-dtmg7t?file=stores%2Fcounter-store.ts)

**Reply by dai-shi** [reply]
> That means I can't read local storage in the initCounterStore ?

If `initCounterStore` is invoked on the server, the server can't read localStorage.
If you limit only to invoke `initCounterStore` on the client, it's fine.
Note that it's not an issue with Zustand, nor even React.

---
Now, there can be some solutions to it. One easy method would be to re-initialize (using localStorage) with `useEffect`.

**Reply by lynn1286** [reply]
> > That means I can't read local storage in the initCounterStore ?
> 
> If `initCounterStore` is invoked on the server, the server can't read localStorage. If you limit only to invoke `initCounterStore` on the client, it's fine. Note that it's not an issue with Zustand, nor even React.
> 
> Now, there can be some solutions to it. One easy method would be to re-initialize (using localStorage) with `useEffect`.

Thanks for the quick reply, I know this question doesn't belong to Zustand or Nextjs or React, I was just wondering where I should get the local storage, because counter-store.ts is not a react component and I can't use useEffect here!

I guess I should read the local store in the useEffect of the component and set my store, right? 🤔

**Reply by dbritto-dev** [reply]
@lynn1286 you need to skipHydration and then hydrate on client. Here you go: https://stackblitz.com/edit/nextjs-and-zustand-persist?file=README.md
